adam-on-the-internet/datamade-code-challenge
# feature/challenge-attempt Adam Callanan

## Overview

This Pull Request recreates DataMade's "Parserator" Web Service functionality. A web form sends a GET request to a Python API, which then uses DataMade's "usaddress" to parse a url and return the parsed response, which the web form then uses to update its view and render the parsed address to the user.

Closes [DataMade's Code Challenge](https://github.com/datamade/code-challenge/blob/master/README.md)

### Demo

Valid Input: 400 Robert D. Ray Drive Des Moines IA 50309

![image](https://user-images.githubusercontent.com/17485463/186986805-220fa4b7-69e4-4d62-9d34-e3cc5d7943d1.png)

Invalid Input: 400 Robert D. Ray Drive Des Moines IA 50309 400 Robert

![image](https://user-images.githubusercontent.com/17485463/186986771-72e5d19c-03d7-410e-9903-7e76f7575670.png)

### Notes

For the purposes of the challenge, in case of a parsing error I return the error's details from usaddress in the REST response and display that message to the user. It would likely be better to provide a less verbose error message to the user.

## Testing Instructions

After checking out this branch locally:

* Setup
  * Install Docker and Docker Compose (restart may be required)
  * Install python requirements with "pip install -r requirements.txt"
  * Build the Docker containers with 'docker-compose build'
  * Run the App with 'docker-compose up' (if you're on Windows, you may need to fix line-endings with 'dos2unix docker-entrypoint.sh')
  * Verify the app is running at http://localhost:8000/
* Test Case: Valid Input
  * Go to the app in your browser: http://localhost:8000/
  * Test valid input by entering "400 Robert D. Ray Drive Des Moines IA 50309" and hitting "parse".
  * Verify the address type and address components show and appear correct.
* Test Case: Invalid Input
  * Go to the app in your browser: http://localhost:8000/
  * Test invalid input by entering "400 Robert D. Ray Drive Des Moines IA 50309 400 Robert" and hitting "parse".
  * Verify a relevant error message displays and that the address type and components do not show.
* Additional Testing
  * If desired, test with more inputs.
  * Run unit tests and linter with 'docker-compose -f docker-compose.yml -f tests/docker-compose.yml run --rm app' (if you're on Windows, you may need to fix line-endings with 'dos2unix scripts/run-tests.sh')
